<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [4]:
### YOUR WORK HERE
import pandas as pd
import numpy as np

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'


column_headers = ['symboling','noromalized-losses','make',
                  'fuel-type','aspirration','num-of-doors','body-style','drive-wheels','engine-location','wheel-base','length','width','height','curb-weight','engine-type','num-of-cylinders','engine-size','fuel-system','bore','stroke','compression-ratio','horsepower','city-mpg','peak-rpm','highway-mpg','price']


df = pd.read_csv(url, names=column_headers)

df = df.replace({"?": np.NaN})


df.head()

symboling noromalized-losses         make  ... peak-rpm highway-mpg  price
0          3                NaN  alfa-romero  ...       21          27  13495
1          3                NaN  alfa-romero  ...       21          27  16500
2          1                NaN  alfa-romero  ...       19          26  16500
3          2                164         audi  ...       24          30  13950
4          2                164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [5]:
df.describe()

symboling  wheel-base  ...    peak-rpm  highway-mpg
count  205.000000  205.000000  ...  205.000000   205.000000
mean     0.834146   98.756585  ...   25.219512    30.751220
std      1.245307    6.021776  ...    6.542142     6.886443
min     -2.000000   86.600000  ...   13.000000    16.000000
25%      0.000000   94.500000  ...   19.000000    25.000000
50%      1.000000   97.000000  ...   24.000000    30.000000
75%      2.000000  102.400000  ...   30.000000    34.000000
max      3.000000  120.900000  ...   49.000000    54.000000

[8 rows x 10 columns]

In [6]:
df.isnull().sum()

symboling              0
noromalized-losses    41
make                   0
fuel-type              0
aspirration            0
num-of-doors           2
body-style             0
drive-wheels           0
engine-location        0
wheel-base             0
length                 0
width                  0
height                 0
curb-weight            0
engine-type            0
num-of-cylinders       0
engine-size            0
fuel-system            0
bore                   4
stroke                 4
compression-ratio      0
horsepower             2
city-mpg               2
peak-rpm               0
highway-mpg            0
price                  4
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [23]:
sample = df.sample(10, random_state=30)
sample

symboling noromalized-losses           make  ... peak-rpm highway-mpg  price
121          1                154       plymouth  ...       31          38   6692
176         -1                 65         toyota  ...       27          32  10898
24           1                148          dodge  ...       31          38   6229
88          -1                137     mitsubishi  ...       23          30   9279
72           3                142  mercedes-benz  ...       16          18  35056
103          0                108         nissan  ...       19          25  13499
26           1                148          dodge  ...       31          38   7609
20           0                 81      chevrolet  ...       38          43   6575
155          0                 91         toyota  ...       27          32   8778
185          2                 94     volkswagen  ...       27          34   8195

[10 rows x 26 columns]

In [8]:
sample.describe()

symboling  wheel-base  ...   peak-rpm  highway-mpg
count  10.000000   10.000000  ...  10.000000    10.000000
mean    0.600000   96.430000  ...  27.000000    32.800000
std     1.264911    2.959373  ...   6.411795     7.269418
min    -1.000000   93.700000  ...  16.000000    18.000000
25%     0.000000   93.900000  ...  24.000000    30.500000
50%     0.500000   96.000000  ...  27.000000    33.000000
75%     1.000000   97.125000  ...  31.000000    38.000000
max     3.000000  102.400000  ...  38.000000    43.000000

[8 rows x 10 columns]

In [9]:
from scipy import stats

In [24]:
stats.ttest_1samp(sample['highway-mpg'], 30.75)

Ttest_1sampResult(statistic=0.8917727404948264, pvalue=0.3957307423595724)

**Due to a T-Statistic of 0.8918 and a P-Value of 0.3957 we fail to reject the null hypothesis. They are rougly the same.**

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [27]:
stats.ttest_1samp(sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

Due to a T-Statistic of -0.9570, pvalue=0.36355472283248624

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [29]:
sample2 = df.sample(100, random_state=30)
stats.ttest_1samp(sample2['highway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

**We can reject the null hypothesis with a sample size of 100. This is because with the larger sample it is significantly more unlikely to have a result that diverges from the supposed mean by such an amount.**

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [31]:
url2 = 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'

column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv(url2, names=column_headers)

print(df.shape)
df.head()


(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [32]:
df = df.replace({'y': 1, 'n': 0, "?": np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [38]:
dem_condition = df['party'] == 'democrat'
dem_df = df[dem_condition]

rep_condition = df['party'] == 'republican'
rep_df = df[rep_condition]

dem_df.head(
)

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [39]:
rep_df.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [42]:
stats.ttest_ind(dem_df['budget'], rep_df['budget'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

Budget issues have significantly more Democrat support.

In [43]:
stats.ttest_ind(dem_df['crime'], rep_df['crime'], nan_policy='omit')

Ttest_indResult(statistic=-16.342085656197696, pvalue=9.952342705606092e-47)

Crime issues have significantly more GOP support.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [15]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [16]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [17]:
### YOUR WORK HERE